### PR TITLE
Wifi basic coop

### PR DESCRIPTION
**Pull request by motwok**
## Description:
Make the wifi component a bit cooperative to other wifi based protocols like ESP-now and ESP-mesh.

Basic means just replace/remove all disabeling of WIFI_STA mode because this removes configuration and disables the other protocol.

To do so the component that needs this simple defines WIFI_BASIC_COOP.

This is mandatory for #1290 to work correctly, due to problems with resources (namely cryptograpy) it is needed that ESP-Now is initialized before Wifi.

**Related issue (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Checklist:
  - [x] The code change is tested and works locally.
  - [-] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [-] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by motwok** [comment]
Push for the first annual

**Comment by d3wy** [comment]
Attempting to compile on latest dev branc comes up the following errors:
```src/esphome/components/network/util.cpp: In function 'esphome::network::IPAddress esphome::network::get_ip_address()':
src/esphome/components/network/util.cpp:36:56: error: could not convert 'esphome::wifi::WiFiComponent::get_ip_address()()' from 'IPAddress' to 'esphome::network::IPAddress'
     return wifi::global_wifi_component->get_ip_address();
                                                        ^
*** [.pioenvs/espnowserver/src/esphome/components/network/util.cpp.o] Error 1
```


Is it possible some other PRs have broken this?

File: esphome/components/wifi/wifi_component_esp32.cpp [hunk 1]
```diff
@@ -467,7 +467,9 @@ void WiFiComponent::wifi_pre_setup_() {
   WiFi.onEvent(f);
   WiFi.persistent(false);
   // Make sure WiFi is in clean state before anything starts
+#ifndef WIFI_BASIC_COOP
```
Comment (martgras) [inline esphome/components/wifi/wifi_component_esp32.cpp:470]: esphome/components/wifi/wifi_component_esp32.cpp no longer exists. 
It was split in wifi_component_esp32.cpp and wifi_component_esp32_arduino

**Review by martgras — CHANGES_REQUESTED** [review]
LGTM in general but wifi_component_esp_idf and wifi_component_esp32_arduino the same modifications. 
Since the PR is older it's based on pre idf refactoring efforts. Needs updating to the new structure

File: esphome/components/wifi/wifi_component_esp32.cpp (hunk 1 above)
Comment (motwok) [reply esphome/components/wifi/wifi_component_esp32.cpp:470]: i see... this will be fixed with a general change soon

**Pull request closed without merging**